amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -111,18 +117,15 @@ public TaskResult call() {
 
                 final long startTime = System.currentTimeMillis();
                 if (reason == ShutdownReason.SHARD_END) {
-                    // Create new lease for the child shards if they don't exist.
-                    // We have one valid scenario that shutdown task got created with SHARD_END reason and an empty list of childShards.
-                    // This would happen when KinesisDataFetcher(for polling mode) or FanOutRecordsPublisher(for StoS mode) catches ResourceNotFound exception.
-                    // In this case, KinesisDataFetcher and FanOutRecordsPublisher will send out SHARD_END signal to trigger a shutdown task with empty list of childShards.
-                    // This scenario could happen when customer deletes the stream while leaving the KCL application running.
+                    final Lease currentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
+
                     if (!CollectionUtils.isNullOrEmpty(childShards)) {
                         createLeasesForChildShardsIfNotExist();
-                        updateLeasesForChildShards();
-                    } else {
-                        log.warn("Shard {} no longer exists. Shutting down consumer with SHARD_END reason without creating leases for child shards.", leaseKeyProvider.apply(shardInfo));
+                        updateLeaseWithChildShards(currentShardLease);
                     }
 
+                    leaseCleanupManager.cleanupLease(currentShardLease);
```

Comment:
1. If we clean up lease here, we should not attempt to checkpoint at applicationCheckpointAndVerification
2. If we don't clean up here and if ddb already has shard end marked, then we should not call applicationCheckpointAndVerification. Basically applicationCheckpointAndVerification should be called only once.

Reply:
1. Yeah good catch
2. Will change so we're only checkpointing if we didn't cleanup the lease and the lease has a non-shard-ended checkpoint.